pkg/sessions/dynamodb/dynamodb_store_test.go

```diff
@@ -0,0 +1,91 @@
+package dynamodb
+
+import (
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	sessionsapi "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/sessions/persistence"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/sessions/tests"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var localDynamo *mockDynamoDBClient
+
+type mockDynamoDBClient struct {
+	dynamodbiface.DynamoDBAPI
+	item map[string]*dynamodb.AttributeValue
+}
+
+func (r *mockDynamoDBClient) DeleteItemWithContext(ctx aws.Context, p *dynamodb.DeleteItemInput, o ...request.Option) (*dynamodb.DeleteItemOutput, error) {
+	for k := range p.Key {
+		delete(r.item, k)
+	}
+
+	return &dynamodb.DeleteItemOutput{}, nil
+}
+
+func (r *mockDynamoDBClient) PutItemWithContext(ctx aws.Context, p *dynamodb.PutItemInput, o ...request.Option) (*dynamodb.PutItemOutput, error) {
+	for k, v := range p.Item {
+		r.item[k] = v
+	}
+	return &dynamodb.PutItemOutput{
+		Attributes:            p.Item,
+		ConsumedCapacity:      nil,
+		ItemCollectionMetrics: nil,
+	}, nil
+}
+
+func (r *mockDynamoDBClient) GetItemWithContext(ctx aws.Context, p *dynamodb.GetItemInput, o ...request.Option) (*dynamodb.GetItemOutput, error) {
+	for k := range p.Key {
+		if r.item[k] != nil {
+			return &dynamodb.GetItemOutput{Item: r.item}, nil
+		}
+	}
+	return &dynamodb.GetItemOutput{}, nil
+}
+
+func (r *mockDynamoDBClient) FastForward(d time.Duration) {
+	now = func() time.Time {
+		return time.Now().Add(d)
```

Comment:
This technically won't make multiple calls to `FastForward` add to eachother, they would each just add duration to `time.Now()`. I think we need to make this aware of previous `FastForwards` each time you add more duration.

We also probably need `now` as a field on the `mockDynamoDBClient` to be used in the methods.

Reply:
What if instead of playing with the `now`, which I think would need to be a part of the store based on where it is being used, a fast forward call instead iterates through all items in the fake dynamo store and subtracts the duration from the expiry of the item? 

That way when it's loaded, the time will look to have been moved forward because the expiry will be x minutes earlier and should cause the validation to fail.

This also has the added benefit of the hackery is in the tests rather than in the product